matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomStateEventsTask.kt

```diff
@@ -0,0 +1,298 @@
+import org.matrix.android.sdk.internal.task.Task
+import org.matrix.android.sdk.internal.util.time.Clock
+import javax.inject.Inject
+
+/**
+ * Generate a list of local state events from the given [CreateRoomBody].
+ * The states events are generated according to the given configuration and following the matrix specification.
+ * This list reflects as much as possible a list of state events related to a real room configured and got from the server.
+ *
+ * Ref: https://spec.matrix.org/latest/client-server-api/#post_matrixclientv3createroom
+ */
+internal interface CreateLocalRoomStateEventsTask : Task<Params, List<Event>> {
+    data class Params(
+            val roomCreatorUserId: String,
+            val createRoomBody: CreateRoomBody
+    )
+}
+
+internal class DefaultCreateLocalRoomStateEventsTask @Inject constructor(
+        private val userService: UserService,
+        private val clock: Clock,
+) : CreateLocalRoomStateEventsTask {
+
+    private lateinit var createRoomBody: CreateRoomBody
+    private lateinit var roomCreatorUserId: String
+
+    override suspend fun execute(params: Params): List<Event> {
+        createRoomBody = params.createRoomBody
+        roomCreatorUserId = params.roomCreatorUserId
+
+        // Build the list of the state events following the priorities from the matrix specification
+        // Changing the order of the events might break the correct display of the room on the client side
+        return buildList {
+            createRoomCreateEvent()
+            createRoomMemberEvents(listOf(roomCreatorUserId))
+            createRoomPowerLevelsEvent()
+            createRoomAliasEvent()
+            createRoomPresetEvents()
+            createRoomInitialStateEvents()
+            createRoomNameAndTopicStateEvents()
+            createRoomMemberEvents(createRoomBody.invitedUserIds.orEmpty())
+            createRoomThreePidEvents()
+            createRoomDefaultEvents()
+        }
+    }
+
+    /**
+     * Generate the create state event related to this room.
+     */
+    private fun MutableList<Event>.createRoomCreateEvent() = apply {
+        val roomCreateEvent = createLocalStateEvent(
+                type = EventType.STATE_ROOM_CREATE,
+                content = RoomCreateContent(
+                        creator = roomCreatorUserId,
+                        roomVersion = createRoomBody.roomVersion,
+                        type = (createRoomBody.creationContent as? Map<*, *>)?.get(CreateRoomParams.CREATION_CONTENT_KEY_ROOM_TYPE) as? String
+
+                ).toContent(),
+        )
+        add(roomCreateEvent)
+    }
+
+    /**
+     * Generate the create state event related to the power levels using the given overridden values or the default values according to the specification.
+     * Ref: https://spec.matrix.org/latest/client-server-api/#mroompower_levels
+     */
+    private fun MutableList<Event>.createRoomPowerLevelsEvent() = apply {
```

Comment:
Using `= apply` is not necessary, since the return value is not used.
Better to use `{`. (at all place in this class)

Reply:
I thought by mistake it was necessary to chain the calls but not for my usage. https://github.com/vector-im/element-android/pull/6127/commits/110cabaca17a3e195279a48ba28cf18bbba0264c